### PR TITLE
Add Spark batch job driver debug helper classes.

### PR DESCRIPTION
**Pull request by wezhang**
The class SparkBatchRemoteDebugJob and SparkBatchDebugSession
are introduced for getting necessary parameters from the
Spark cluster and creating SSH port forwarding tunnel.

The code is no referred now. And it will be used in Spark Job
IDE runner and executor.

Unit tests and integration tests with mocked http server
are included.

Signed-off-by: Zhang Wei <zhwe@microsoft.com>

### CODE REVIEW

**Comment by vscjavaci** [comment]
Can one of the admins verify this patch?

**Comment by wezhang** [comment]
Sorry for a lot of new codes accumulated in the migration.

**Comment by wezhang** [comment]
Hit in Travis CI:

```
129.73s$ mvn install -f PluginsAndFeatures/azure-toolkit-for-eclipse/pom.xml -q
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=192m; support was removed in 8.0
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-packaging-plugin:0.25.0:package-plugin (default-package-plugin) on project com.microsoft.azuretools.sdk: /home/travis/build/Microsoft/azure-tools-for-java/PluginsAndFeatures/AddLibrary/AzureLibraries/com.microsoft.azuretools.sdk/build.properties: bin.includes value(s) [dependencies/slf4j-api-1.7.22.jar] do not match any files. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :com.microsoft.azuretools.sdk
The command "mvn install -f PluginsAndFeatures/azure-toolkit-for-eclipse/pom.xml -q" exited with 1.
```

File: Utils/hdinsight-node-common/Test/resources/log4j.properties [hunk 1]
```diff
@@ -0,0 +1,5 @@
+log4j.rootLogger=ERROR, stdout
+log4j.appender.stdout=org.apache.log4j.ConsoleAppender
+log4j.appender.stdout.Target=System.out
+log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
+log4j.appender.stdout.layout.ConversionPattern=%d{yyyy-MM-dd'T'HH:mm:ss.SSS} %-5p [%c] - %m%n
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/Test/resources/log4j.properties]: Should we have such a logger properties file?

Comment (wezhang) [inline Utils/hdinsight-node-common/Test/resources/log4j.properties]: It's for log control in testing

**Comment by vscjavaci** [comment]
Can one of the admins verify this patch?

**Comment by voidfunction** [comment]
Can one of the admins verify this patch?

File: Utils/hdinsight-node-common/pom.xml [hunk 2]
```diff
@@ -189,6 +189,11 @@
             <artifactId>httpclient</artifactId>
         </dependency>
         <dependency>
+            <groupId>com.jcraft</groupId>
+            <artifactId>jsch</artifactId>
+            <version>0.1.54</version>
+        </dependency>
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/pom.xml]: remove the version from this pom and add it to dependencysmanagement in Utils/pom.xml

**Comment by voidfunction** [comment]
>             <version>1.9.1</version>

please help remove the version information too.

---
Refers to: Utils/hdinsight-node-common/pom.xml:150 in 3c3d877. [](commit_id = 3c3d8775246ae3b88dbcac2fd678712b2945a088, deletion_comment = False)

File: Utils/hdinsight-node-common/pom.xml [hunk 3]
```diff
@@ -189,6 +189,11 @@
             <artifactId>httpclient</artifactId>
         </dependency>
         <dependency>
+            <groupId>com.jcraft</groupId>
+            <artifactId>jsch</artifactId>
+            <version>0.1.54</version>
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/pom.xml]: ><version>0.1.54</version> [](start = 12, length = 25)

same

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ILogger.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * Base logger class
+ */
+public interface ILogger {
+    default org.slf4j.Logger log() {
+        return LoggerFactory.getLogger(getClass());
+    }
+}
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ILogger.java]: DefaultLoader.getUIHelper().logError(); we have such an interfact to logger error in IntelliJ and Eclipse. I suppose we can keep the same style if we are willing to save the log?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkJobException.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * SOFTWARE.
+ *
+ */
+
+package com.microsoft.azure.hdinsight.spark.common;
+
+/**
+ * The Base Exception for all Spark Job related exceptions
+ */
+public class SparkJobException extends Exception{
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkJobException.java]: >Exception{ [](start = 39, length = 10)

HDIException for consistency?

File: Utils/hdinsight-node-common/pom.xml [hunk 6]
```diff
@@ -172,6 +172,11 @@
             <version>1.2.17</version>
         </dependency>
         <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-log4j12</artifactId>
+            <version>1.7.22</version>
+        </dependency>
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/pom.xml]: remove the version information.

File: Utils/hdinsight-node-common/pom.xml [hunk 7]
```diff
@@ -172,6 +172,11 @@
             <version>1.2.17</version>
         </dependency>
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/pom.xml]: this dependency should be removed if slf4j-log412 is added to this project.

File: Utils/hdinsight-node-common/pom.xml (hunk 2 above)
Comment (wezhang) [inline Utils/hdinsight-node-common/pom.xml]: Fix it.

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkConfigures.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.microsoft.azure.hdinsight.spark.common;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class SparkConfigures extends HashMap<String, Object> {
+    public SparkConfigures(Map<? extends String, ?> m) {
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkConfigures.java]: >? extends String [](start = 31, length = 16)

which means String type can't be the key?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkJobException.java (hunk 5 above)
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkJobException.java]: Fix it.

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ILogger.java (hunk 4 above)
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ILogger.java]: Fix it by adding UILoggerAppender for log4j

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkConfigures.java (hunk 8 above)
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkConfigures.java]: Fix it

File: Utils/hdinsight-node-common/pom.xml (hunk 6 above)
Comment (wezhang) [inline Utils/hdinsight-node-common/pom.xml]: Fix it!

File: Utils/hdinsight-node-common/pom.xml (hunk 7 above)
Comment (wezhang) [inline Utils/hdinsight-node-common/pom.xml]: Fix it!

File: Utils/hdinsight-node-common/pom.xml (hunk 3 above)
Comment (wezhang) [inline Utils/hdinsight-node-common/pom.xml]: Fix it!

**Comment by voidfunction** [comment]
`171.24s$ mvn install -f PluginsAndFeatures/azure-toolkit-for-eclipse/pom.xml -q
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=192m; support was removed in 8.0
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-packaging-plugin:0.25.0:package-plugin (default-package-plugin) on project com.microsoft.azuretools.sdk: /home/travis/build/Microsoft/azure-tools-for-java/PluginsAndFeatures/AddLibrary/AzureLibraries/com.microsoft.azuretools.sdk/build.properties: bin.includes value(s) [dependencies/joda-time-2.4.jar] do not match any files. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :com.microsoft.azuretools.sdk
`

**Comment by wezhang** [comment]
Thanks! Fixed.

**Comment by vscjavaci** [comment]
Can one of the admins verify this patch?

**Comment by voidfunction** [comment]
Can one of the admins verify this patch?

**Comment by voidfunction** [comment]
ok to test

**Pull request merged**